File: IT/Multithreading/Python/copyshop.py
# file copyshop.py

from __future__ import print_function
from multiprocessing import Process, current_process, Semaphore, Lock
from random import randint
from time import sleep

def printingJob(document, printer, lock):
	# documento inviato in un intervallo di tempo tra 1 e 25 secondi
	pause = randint(1, 25)
	sleep(pause)

	with lock:
		print("{} invia il documento \"{}\" alla stampante.".format(current_process().name, document))

	# acquisisce il semaforo-stampante
	printer.acquire()

	pause = randint(5, 10)
	sleep(pause)
	print("Stampato il documento \"{}\" inviato dal {}.".format(document, current_process().name))

	# libera il semaforo-stampante
	printer.release()

if __name__ == "__main__":
	# il semaforo rappresenta una stampante condivisa
	printer = Semaphore(1)
	lock = Lock()

	documents = ["Modulo di iscrizione", "Appunti d'esame", "Tesi laurea triennale", "Tesi laurea specialistica", "Tesi dottorato"]

	students = [Process(target=printingJob, args=(documents[randint(0, len(documents)-1)], printer, lock,)) for i in range(10)]

	for student in students:
		student.start()

	for student in students:
		student.join()

File: IT/Multithreading/Java/CopyShop.java
/* file CopyShop.java */

import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.Semaphore;

public class CopyShop {
	private static class PrintingJob implements Runnable {
		private String document;
		private Semaphore printer;

		public PrintingJob(String document, Semaphore printer) {
			this.document = document;
			this.printer = printer;
		}

		@Override
		public void run() {
			try {
				// documento inviato in un intervallo di tempo tra 1 e 25 secondi
				int pause = ThreadLocalRandom.current().nextInt(1, 25+1);
				Thread current = Thread.currentThread();
				current.sleep(pause*1000);

				System.out.println(current.getName() + " invia il documento \"" + document + "\" alla stampante.");

				// acquisisce il semaforo-stampante
				printer.acquire();

				pause = ThreadLocalRandom.current().nextInt(5, 10+1);
				current.sleep(pause*1000);
				System.out.println("Stampato il documento \"" + document + "\" inviato dal " + current.getName() + ".");

				// libera il semaforo-stampante
				printer.release();
			}
			catch (InterruptedException e) {
				System.err.println(e);
			}
		}
	}

	public static void main(String [] args) {
		// il semaforo rappresenta una stampante condivisa
		Semaphore printer = new Semaphore(1);

		String [] documents = { "Modulo di iscrizione", "Appunti d'esame", "Tesi laurea triennale", "Tesi laurea specialistica", "Tesi dottorato" };

		for (int i = 0; i < 10; i++) {
			int t = ThreadLocalRandom.current().nextInt(documents.length);
			Thread student = new Thread(new PrintingJob(documents[t], printer));
			student.start();
		}
	}
}

